compat/mingw.c

```diff
@@ -1097,6 +1110,20 @@ int mingw_lstat(const char *file_name, struct stat *buf)
 		filetime_to_timespec(&(fdata.ftLastAccessTime), &(buf->st_atim));
 		filetime_to_timespec(&(fdata.ftLastWriteTime), &(buf->st_mtim));
 		filetime_to_timespec(&(fdata.ftCreationTime), &(buf->st_ctim));
+		if (wsl_compat && !S_ISLNK(buf->st_mode) && !S_ISDIR(buf->st_mode)) {
+			HANDLE hf = CreateFileW(
+				wfilename, FILE_READ_EA | SYNCHRONIZE,
+				FILE_SHARE_READ | FILE_SHARE_WRITE |
+					FILE_SHARE_DELETE,
+				NULL, OPEN_EXISTING,
+				FILE_FLAG_BACKUP_SEMANTICS |
+					FILE_FLAG_OPEN_REPARSE_POINT,
+				NULL);
+			if (hf != INVALID_HANDLE_VALUE) {
+				read_wsl_stat(hf, buf);
```

Comment:
In the same vein as my comment about `write_wsl_stat()`, how about renaming this function to `copy_wsl_mode_bits_from_disk()`?

There are three callers of this function, two of which only have a file name in wide characters to work with. My preference, therefore, would be to extend the signature of that function so that it takes a `HANDLE` _or_ a `wchar_t *`. Something like this:

```c
int copy_wsl_mode_bits_from_disk(HANDLE h, wchar_t *wfilename, struct stat *st)
{
  [...]

  assert(!!h + !!wfilename == 1);
  if (!h)
    h = CreateFileW(wfilename, FILE_READ_EA | SYNCHRONIZE,
		    FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
		    NULL, OPEN_EXISTING,
		    FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT,
		    NULL);
```

This is needed also if we want to encapsulate the `wsl_enabled` check properly in `compat/win32/wsl.c`.

Oh and BTW there are plenty of places where a `HANDLE` is named `h` , `hnd` or `handle`. That's already pretty inconsistent as it is, let's not introduce yet another name (`hf`).